File: src/types/parcel.ts
export type FeatureCategory = 'Building' | 'Road' | 'Vegetation' | 'Fence' | 'Driveway' | 'Pool' | 'Shed' | 'Parcel' | 'Verified Boundary';

export type MapMode = 'select' | 'pin' | 'draw' | 'measure_distance' | 'measure_area';

export type MapLayerType = 
  | 'street' 
  | 'google_street' 
  | 'google_satellite' 
  | 'google_hybrid' 
  | 'google_terrain' 
  | 'satellite' 
  | 'terrain';

export interface GeoJSONGeometry {
  type: 'Polygon';
  coordinates: number[][][]; // [[lng, lat], ...]
}

export interface Parcel {
  id: string;
  name: string;
  area_m2: number;
  area_sqft: number;
  area_acres: number;
  area_hectares: number;
  perimeter_m: number;
  confidence: number;
  features: string[];
  latitude: number;
  longitude: number;
  status: 'User Created' | 'AI Detected' | 'Validated';
  geometry: GeoJSONGeometry;
}

export interface PinLocation {
  id: string;
  latitude: number;
  longitude: number;
  title?: string;
  createdAt: string;
}

export interface MeasurementResult {
  id: string;
  type: 'distance' | 'area';
  points: [number, number][]; // [lat, lng]
  distance_m?: number;
  area_m2?: number;
  area_sqft?: number;
  area_acres?: number;
  area_hectares?: number;
}

export interface FeatureLayerVisibility {
  parcels: boolean;
  buildings: boolean;
  roads: boolean;
  vegetation: boolean;
  verified: boolean;
}

export interface ProcessingStep {
  id: number;
  title: string;
  status: 'pending' | 'processing' | 'completed' | 'error';
  log: string;
}

export interface LogEntry {
  id: string;
  timestamp: string;
  text: string;
  type: 'info' | 'processing' | 'success' | 'error';
}

export interface BackendStatus {
  online: boolean;
  service?: string;
  message?: string;
}

File: backend/app/schemas/parcel.py
from pydantic import BaseModel, Field
from typing import List, Optional, Any, Dict

class GeoJSONGeometry(BaseModel):
    type: str = "Polygon"
    coordinates: List[List[List[float]]]

class FeatureSchema(BaseModel):
    id: str
    parcel_id: str
    feature_type: str  # Building, Road, Vegetation, Fence, Driveway, Pool, Shed
    confidence: float
    geometry: Optional[GeoJSONGeometry] = None

class ParcelBase(BaseModel):
    name: str = "Parcel"
    area_m2: float
    area_sqft: float
    area_acres: float
    area_hectares: float
    perimeter_m: float
    confidence: float = 100.0
    features: List[str] = Field(default_factory=list)
    latitude: float
    longitude: float
    status: str = "User Created"  # User Created, AI Detected, Validated

class ParcelCreate(ParcelBase):
    id: Optional[str] = None
    geometry: GeoJSONGeometry

class ParcelUpdate(BaseModel):
    name: Optional[str] = None
    status: Optional[str] = None
    geometry: Optional[GeoJSONGeometry] = None
    features: Optional[List[str]] = None
    area_m2: Optional[float] = None
    area_sqft: Optional[float] = None
    area_acres: Optional[float] = None
    area_hectares: Optional[float] = None
    perimeter_m: Optional[float] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None

class ParcelResponse(ParcelBase):
    id: str
    geometry: GeoJSONGeometry

    class Config:
        from_attributes = True

class FeatureCollectionGeoJSON(BaseModel):
    type: str = "FeatureCollection"
    features: List[Dict[str, Any]]
